src/server/mod.rs

```diff
@@ -0,0 +1,321 @@
+
+                tokio::spawn(async move {
+                    debug!("Handling connection");
+                    if let Err(e) = handle_conn(conn, block_store_readonly, best_chain, ledger, summary).await {
+                        error!("Error handling connection: {e}");
+                    }
+                    debug!("Removing readonly instance at {}", secondary_path.display());
+                    tokio::fs::remove_dir_all(&secondary_path).await.ok();
+                });
+            }
+        }
+    }
+}
+
+async fn handle_conn(
+    conn: LocalSocketStream,
+    db: BlockStoreConn,
+    best_chain: Vec<BlockHash>,
+    ledger: Ledger,
+    summary: Summary,
+) -> Result<(), anyhow::Error> {
+    let (reader, mut writer) = conn.into_split();
+    let mut reader = BufReader::new(reader);
+    let mut buffer = Vec::with_capacity(128);
+    let _read = reader.read_until(0, &mut buffer).await?;
+
+    let mut buffers = buffer.split(|byte| *byte == 32);
+    let command = buffers.next().unwrap();
+
+    let command_string = String::from_utf8(command.to_vec())?;
+
+    match command_string.as_str() {
+        "account" => {
+            let data_buffer = buffers.next().unwrap();
+            let public_key = PublicKey::from_address(&String::from_utf8(
+                data_buffer[..data_buffer.len() - 1].to_vec(),
+            )?)?;
+            info!("Received account command for {public_key:?}");
+            debug!("Using ledger {ledger:?}");
+            let account = ledger.accounts.get(&public_key);
+            if let Some(account) = account {
+                debug!("Writing account {account:?} to client");
+                let bytes = bcs::to_bytes(account)?;
+                writer.write_all(&bytes).await?;
+            }
+        }
+        "best_chain" => {
+            info!("Received best_chain command");
+            let data_buffer = buffers.next().unwrap();
+            let num = String::from_utf8(data_buffer[..data_buffer.len() - 1].to_vec())?
+                .parse::<usize>()?;
+            let best_chain: Vec<PrecomputedBlock> = best_chain[..best_chain.len() - 1]
+                .iter()
+                .take(num)
+                .cloned()
+                .map(|state_hash| db.get_block(&state_hash.0).unwrap().unwrap())
+                .collect();
+            let bytes = bcs::to_bytes(&best_chain)?;
+            writer.write_all(&bytes).await?;
+        }
+        "best_ledger" => {
+            info!("Received best_ledger command");
+            let data_buffer = buffers.next().unwrap();
+            let path = &String::from_utf8(data_buffer[..data_buffer.len() - 1].to_vec())?
+                .parse::<PathBuf>()?;
+            debug!("Writing ledger to {}", path.display());
+            fs::write(path, format!("{ledger:?}")).await?;
+            let bytes = bcs::to_bytes(&format!("Ledger written to {}", path.display()))?;
+            writer.write_all(&bytes).await?;
+        }
+        "summary\0" => {
```

Comment:
why does this one still have a `\0` when the others don't? do we need it or not?

Reply:
It needs it because it doesn't have an y accompanying args

Reply:
Without it, client hangs